Watershader improvements

Astrak opened the pull request:
This PR first changes the settings of the shader_ocean example :
- reduced scene scale (there was a lot of zeros, making it difficult to understand what is at which scale, and difficult to edit)
- added dat gui (to play with the watershader parameters)

It also improves the watershader :
- added the _size_ parameter
- supports shadows
- supports toneMapping

Finally, it also changes the API : 

- instead of creating the water object, then the mirror mesh, appending the former to the latter, and using the object's material... this is hidden in `THREE.WaterMesh`, and we can now just call `water = new THREE.WaterMesh()`.

There still is the weird reflection bias to fix, not sure how to do it. But the icosahedron is now doublesided so it's not visible anymore.

Mugen87 commented:
For testing: http://rawgit.com/Astrak/three.js/watershader-improvements/examples/webgl_shaders_ocean.html

Mugen87 commented on examples/webgl_shaders_ocean.html:
```diff
@@ -80,29 +81,26 @@
 				controls = new THREE.OrbitControls( camera, renderer.domElement );
-				controls.enablePan = false;
-				controls.minDistance = 1000.0;
-				controls.maxDistance = 5000.0;
```
I would not delete these lines. If the user zooms too far out, the scene looks weird. Same for zooming in...

Astrak replied:
i re-inserted them in the latest commits, with new values since the scene scale is different.

Mugen87 commented:
Your implementation is okay, i just have some concerns about the missing backwards compatibility of your change. When users update to a new version of `three.js`, their existing code will definitely break. So this PR needs to be noted in the migration guide if it will be merged.

Mugen87 commented on examples/js/WaterShader.js:
```diff
@@ -140,16 +140,20 @@ THREE.Water = function ( renderer, camera, scene, options ) {
-			'	vec3 worldToEye = eye-worldPosition;',
+			'	vec3 worldToEye = eye-worldPosition.xyz;',
 			'	vec3 eyeDirection = normalize( worldToEye );',
 			'	sunLight( surfaceNormal, eyeDirection, 100.0, 2.0, 0.5, diffuseLight, specularLight );',
```
Does it still make sense to support `sunDirection` and `sunColor` when the material reacts on light source now?

Astrak replied:
I think it does, the bsdfs are still not used except the one of the shader, that gives his look. The new chunks are only here to support the shadowmap. However your question made me check with different lights, and the shader is currently broken if we use other types of lights than the directional lights. Working on it.

Astrak commented:
How to edit the future migration guide from r87 to r88 ?

Astrak commented:
The light issue is fixed now, but you can check that the lighting is still the one of the shader, it doesn't use the other provided lights, so we still need these uniforms. The main goal of this PR is just to add a native support of shadows into `THREE.Water` so we can call `water.receiveShadow = true` as any other mesh. I did not intent to change the lighting implementation here.

How does one update the migration guide ?

Astrak commented:
This commit should ensure backwards compatibility : `THREE.Water` is unchanged. The new water, added on top of it with the nested objects pattern, is called `THREE.WaterMesh`.

Mugen87 commented:
Okay. Thanks for clarification!

Astrak commented:
Maybe i could ping the original author mentionned in WaterShader.js @jbouny, to check this PR and have an advice on that reflection bias : I'm sure I had it solved in a sketch I made some time ago by reducing the scene near plane, but it doesn't change anything here, though it only is a problem at small scale... very weird to me.

Astrak commented:
i think i just really made awefull things with git...

Astrak committed:
changes back

reset head after local merge of dev and watershader-improvements and head reset...

mrdoob commented:
While you're at it... How about making this simpler to use by adopting `onBeforeRender`?

https://github.com/mrdoob/three.js/blob/dev/examples/js/Mirror.js#L116-L240

That way we don't need to pass `renderer`, `camera` nor `scene` in the constructor.

Astrak committed:
onBeforeRender-ready but not ok

removed renderTarget2 (was not used)
implemented onBeforeRender, but seems that the shadow rendering creates an infinite loop of onBeforeRender-s, so just renamed it 'render' for now, before working on that, new pattern ok.
generates a warning 'source and destination textures are the same', I don't understand why for now.

Astrak committed:
onBeforeRender ok

mesh.visible instead of material.visible... that was the issue. The rest also is now similar to the mirror render chunk inside its onBeforeRender callback.

mrdoob commented on examples/js/WaterShader.js:
```diff
@@ -162,10 +151,11 @@ THREE.Water = function ( renderer, camera, scene, options ) {
 			'	float reflectance = rf0 + ( 1.0 - rf0 ) * pow( ( 1.0 - theta ), 5.0 );',
 			'	vec3 scatter = max( 0.0, dot( surfaceNormal, eyeDirection ) ) * waterColor;',
-			'	vec3 albedo = mix( sunColor * diffuseLight * 0.3 + scatter, ( vec3( 0.1 ) + reflectionSample * 0.9 + reflectionSample * specularLight ), reflectance );',
+			'	vec3 albedo = mix( ( sunColor * diffuseLight * 0.3 + scatter ) * getShadowMask(), ( vec3( 0.1 ) + reflectionSample * 0.9 + reflectionSample * specularLight ), reflectance);',
```
Any way the shadow could get affected by the displacement too? Right now feels like a mirror.

mrdoob replied:
I'll merge in the meantime though 😊

mrdoob commented:
Thanks!
